### PR TITLE
Use an interleaved tags-pointers-tags-... layout for hash tables

### PR DESCRIPTION
**Pull request by oerling**
Use an interleaved tags-pointers-tags-... layout for hash tables

Organizes join and group by hash tables into chunks of 2 cache lines
with 16 tag bytes followed by 16 48 bit pointers. This is analogous to
F14 layout where groups of 16 8 bit tags alternate with payload.

In principle, for hash tables that miss often and where the miss can
be detected with tags alone, putting the tags all together could be
better for locality. Experimentally, we do not see much difference, in
any case not enough to maintain two hash table layouts. On the other
hand, when interleaving tags and pointers, we quite often save a cache
miss for a hit because the tag is in the same cache line as the
pointer. This is seen to be ~15% faster than a layout with separate
tags and pointers. See hashTableBenchmark.cpp for probe level clock
counts with different table sizes and hit rates.

The hash table arrays are slightly smaller, at 8 bytes per entry
instead of 9 bytes wen the tags were separate.
Comparison vs F14 (HashTableBenchmark.cpp on core i7 laptop)

We show the count of entries, and the time to probe in clocks for
HashTable.cpp and F14. The percentage is the percentage of time F14
takes over the Velox time.  > 100 means F14 is slower).

Hit4M: Rows=4000000 Hit%=100 NumProbes=4000000 hash/row=4.22324 probe clocks=80.4019 f14Probe=107.357 (133.526%)
 numDistinct=4000000 mode=normalized key
Miss4M: Rows=4000000 Hit%=5 NumProbes=80000000 hash/row=3.09183 probe clocks=53.8863 f14Probe=76.7055 (142.347%)
 numDistinct=4802527 mode=normalized key
Hit32M: Rows=32000000 Hit%=100 NumProbes=32000000 hash/row=3.92055 probe clocks=86.266 f14Probe=116.758 (135.347%)
 numDistinct=32000000 mode=normalized key
Miss32M: Rows=32000000 Hit%=5 NumProbes=640000000 hash/row=3.18426 probe clocks=83.8733 f14Probe=97.35 (116.068%)
 numDistinct=38398149 mode=normalized key
Hit128M: Rows=128000000 Hit%=100 NumProbes=128000000 hash/row=3.15532 probe clocks=121.393 f14Probe=184.387 (151.894%)
 numDistinct=128000000 mode=normalized key

### CODE REVIEW

File: velox/exec/HashTable.cpp [hunk 1]
```diff
@@ -26,6 +26,8 @@
 
 using facebook::velox::common::testutil::TestValue;
 
+DEFINE_bool(no_unroll, false, "No interleave of probe cache misses");
```
Comment (mbasmanova) [inline velox/exec/HashTable.cpp:29]: As discussed offline, let's remove this flag.

File: velox/exec/HashTable.cpp [hunk 2]
```diff
@@ -290,6 +294,21 @@ void HashTable<ignoreNullKeys>::storeRowPointer(
...
+    // a 6 byte pointer to the row.
+    int groupOffset = index & ~(kTagRowGroupSize - 1);
+    uint64_t* pointer = reinterpret_cast<uint64_t*>(
+        tags_ + sizeof(TagVector) + groupOffset +
+        (kBytesInPointer * (index - groupOffset)));
+
+    // We store 48 bits, preserving the high 16 bits of the word, which belong
+    // to the next pointer.
+    auto previous = *pointer & ~kPointerMask;
+    *pointer = reinterpret_cast<uint64_t>(row) | previous;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:310]: How about
```
#define SET_ROW_POINTER(pointer, row) \
*pointer = (*pointer & ~kPointerMask) | (kPointerMask & reinterpret_cast<uint64_t>(row))
```

File: velox/exec/HashTable.cpp [hunk 3]
```diff
@@ -420,6 +435,10 @@ void populateNormalizedKeys(HashLookup& lookup, int8_t sizeBits) {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::groupProbe(HashLookup& lookup) {
+  if (kTrackLoads) {
+    numProbe_ += lookup.rows.size();
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:439]: Can we define a method for this as the other stats collection? Thanks!

File: velox/exec/HashTable.cpp [hunk 4]
```diff
@@ -516,12 +596,11 @@ void HashTable<ignoreNullKeys>::arrayGroupProbe(HashLookup& lookup) {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::joinProbe(HashLookup& lookup) {
+  if (kTrackLoads) {
+    numProbe_ += lookup.rows.size();
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:600]: ditto

File: velox/exec/HashTable.h [hunk 5]
```diff
@@ -679,6 +740,10 @@ class HashTable : public BaseHashTable {
   // many threads can set this.
   std::atomic<bool> hasDuplicates_{false};
 
+  // True if tombstones need to be checked. An erase from a group with no
+  // empties makes a tombstone.
+  bool hasTombstones_{false};
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:745]: Do we need this as we already have numTombstones_ tracked?

File: velox/exec/HashTable.h [hunk 6]
```diff
@@ -668,6 +685,50 @@ class HashTable : public BaseHashTable {
     return isJoinBuild_ ? 0 : 50;
   }
 
+  // Returns the offset in bytes of the tag word for 'hash'. The offset is
+  // from 'tags_'.
+  int32_t tagVectorOffset(uint64_t hash) const {
+    return hash & tagOffsetMask_;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:691]:
```
s/tagOffsetMask_/tagGroupOffsetMask/?
```

File: velox/exec/HashTable.h [hunk 7]
```diff
@@ -696,6 +766,21 @@ class HashTable : public BaseHashTable {
   // Owns the memory of multiple build side hash join tables that are
   // combined into a single probe hash table.
   std::vector<std::unique_ptr<HashTable<ignoreNullKeys>>> otherTables_;
+  // Statistics maintained if kTrackCollisions is set.
+
+  // Number of times a row is looked up or inserted.
+  mutable tsan_atomic<int64_t> numProbe_{0};
+
+  // Number of times a word of 16 tags is accessed. at least once per probe.
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:774]:
```
s/accessed. at/accessed with at/
```

File: velox/exec/HashTable.cpp [hunk 8]
```diff
@@ -290,6 +294,21 @@ void HashTable<ignoreNullKeys>::storeRowPointer(
...
   if (hashMode_ != HashMode::kArray) {
     tags_[index] = hashTag(hash);
+    // The pointer is in slot (index - start_of_group) after the
+    // tags. We first get the base address of the tag/pointer
+    // group. We add the size of the tags. Then we add pointer size
+    // * index of the tag in the tags vector. This is the address of
+    // a 6 byte pointer to the row.
+    int groupOffset = index & ~(kTagRowGroupSize - 1);
+    uint64_t* pointer = reinterpret_cast<uint64_t*>(
+        tags_ + sizeof(TagVector) + groupOffset +
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:304]:
```
tags_ + groupOffset + sizeof(TagVector) + ..
```

File: velox/exec/HashTable.cpp [hunk 9]
```diff
@@ -777,18 +896,21 @@ void HashTable<ignoreNullKeys>::parallelJoinBuild() {
       buildPartitionBounds_.begin(),
       buildPartitionBounds_.begin() + buildPartitionBounds_.capacity(),
       std::numeric_limits<PartitionBoundIndexType>::max());
+
+  // The partitioning is in terms of ranges of tag vector index.
+  int64_t tagIndexEnd = sizeMask_ + 1;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:901]: tagIndexEnd is not used?

File: velox/exec/HashTable.cpp [hunk 10]
```diff
@@ -777,18 +896,21 @@ void HashTable<ignoreNullKeys>::parallelJoinBuild() {
...
+    buildPartitionBounds_[i] =
+        bits::roundUp(((sizeMask_ + 1) / numPartitions) * i, kTagRowGroupSize);
     // Bounds must always be positive
     VELOX_CHECK_GE(
         buildPartitionBounds_[i],
         0,
         "Turn on VELOX_ENABLE_INT64_BUILD_PARTITION_BOUND to avoid integer overflow in buildPartitionBounds_");
   }
-  buildPartitionBounds_.back() = capacity_;
+  buildPartitionBounds_.back() = sizeMask_ + 1;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:913]: Why need this change?

File: velox/exec/HashTable.cpp [hunk 11]
```diff
@@ -848,7 +970,7 @@ void HashTable<ignoreNullKeys>::parallelJoinBuild() {
         hashes.data(),
         overflows.size(),
         0,
-        capacity_,
+        sizeMask_ + 1,
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:973]: ditto

File: velox/exec/HashTable.cpp [hunk 12]
```diff
@@ -962,8 +1084,13 @@ void HashTable<ignoreNullKeys>::insertForGroupBy(
   } else {
     for (int32_t i = 0; i < numGroups; ++i) {
       auto hash = hashes[i];
-      auto tagIndex = ProbeState::tagsByteOffset(hash, sizeMask_);
+      auto tagIndex = tagVectorOffset(hash);
       auto tagsInTable = BaseHashTable::loadTags(tags_, tagIndex);
+      if (i + 6 < numGroups) {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:1089]: Can we do this in followup? Thanks!

File: velox/exec/HashTable.h [hunk 13]
```diff
@@ -491,6 +485,67 @@ class HashTable : public BaseHashTable {
   }
 
  private:
+  // Enables debug stats for collisions. Should be false for normal operation.
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:488]:
```
// Enables debug stats for collisions on debug build.
```

File: velox/exec/HashTable.h [hunk 14]
```diff
@@ -682,11 +787,19 @@ class HashTable : public BaseHashTable {
...
   int32_t nextOffset_;
-  uint8_t* tags_ = nullptr;
   char** table_ = nullptr;
   memory::ContiguousAllocation tableAllocation_;
+
+  // Number of slots across all buckets.
   int64_t capacity_{0};
+
+  // Mask for extracting low bits of hash number for use as byte offsets into
+  // the table. This is (capacity_ * sizeof(void*) - 1.
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:797]:
```
This is set to capacity_ * sizeof(void*) - 1.
```

File: velox/exec/HashTable.h [hunk 15]
```diff
@@ -668,6 +729,50 @@ class HashTable : public BaseHashTable {
     return isJoinBuild_ ? 0 : 50;
   }
 
+  // Returns the offset in bytes of the bucket for 'hash'.
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:732]:
```
// Returns the byte offset of the bucket for 'hash' starting from 'table_'.
```

File: velox/exec/HashTable.h [hunk 16]
```diff
@@ -668,6 +729,50 @@ class HashTable : public BaseHashTable {
...
+  int32_t bucketOffset(uint64_t hash) const {
+    return hash & bucketOffsetMask_;
+  }
+
+  // Returns the offset of the next bucket from 'offset'. Wraps around at
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:737]:
```
Returns the byte offset
```

File: velox/exec/HashTable.h [hunk 17]
```diff
@@ -668,6 +729,50 @@ class HashTable : public BaseHashTable {
...
+  // the end of the table.
+  int32_t nextBucketOffset(int32_t offset) const {
+    return sizeMask_ & (offset + kBucketSize);
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:740]: Can you add a check
```
VELOX_DCHECK_EQ(offset & (kBucketSize - 1), 0);
```

File: velox/exec/HashTable.h [hunk 18]
```diff
@@ -668,6 +729,50 @@ class HashTable : public BaseHashTable {
...
+  }
+
+  char* row(int64_t bucketOffset, int32_t slotOffset) const {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:743]:
```
// Returns the row pointer of the specified slot from a given bucket.
s/slotOffset/slotIndex/
```

File: velox/exec/HashTable.h [hunk 19]
```diff
@@ -668,6 +729,50 @@ class HashTable : public BaseHashTable {
...
+    return bucketAt(bucketOffset)->pointerAt(slotOffset);
+  }
+
+  TagVector loadTags(int32_t bucketOffset) const {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:747]:
```
// Loads the tag vector of a given bucket.
```

File: velox/exec/HashTable.h [hunk 20]
```diff
@@ -491,6 +485,67 @@ class HashTable : public BaseHashTable {
...
+  static constexpr bool kTrackLoads = false;
+#else
+  static constexpr bool kTrackLoads = true;
+#endif
+
+  static constexpr uint8_t kPointerSignificantBits = 48;
+  static constexpr uint64_t kPointerMask =
+      bits::lowMask(kPointerSignificantBits);
+
+  // lines.The table has a power of two
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:499]:
```
// The table in non-array mode has ...
```

File: velox/exec/HashTable.h [hunk 21]
```diff
@@ -491,6 +485,67 @@ class HashTable : public BaseHashTable {
...
+  // number of buckets with 16
+  // slots. Each slot has a 1 byte tag
+  // (a field of hash number) and a 48
+  // bit pointer. All the tags are in a
+  // 16 byte SIMD word,, followed by the
+  // 6 byte pointers. There are 16 bytes
+  // of padding at the end to make the
+  // bucket occupy exactly two cache lies.
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:507]: Here we assume each hw cache line is 64 bytes?

File: velox/exec/HashTable.h [hunk 22]
```diff
@@ -491,6 +485,67 @@ class HashTable : public BaseHashTable {
...
+          *reinterpret_cast<uintptr_t*>(&pointers[kPointerSize * slotIndex]) &
+          kPointerMask);
+    }
+
+    void setTag(int32_t slotIndex, uint8_t tag) {
+      reinterpret_cast<uint8_t*>(&tags)[slotIndex] = tag;
+    }
+
+    void setPointer(int32_t slotIndex, void* pointer) {
+      const auto slot =
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:532]:
```
const auto* slot =
```

File: velox/exec/HashTable.h [hunk 23]
```diff
@@ -491,6 +485,67 @@ class HashTable : public BaseHashTable {
...
 
  private:
+  // Enables debug stats for collisions. Should be false for normal operation.
+#ifdef NDEBUG
+  static constexpr bool kTrackLoads = false;
+#else
+  static constexpr bool kTrackLoads = true;
+#endif
+
+  static constexpr uint8_t kPointerSignificantBits = 48;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:495]: Can we move these two consts as private to Bucket? Thanks!

File: velox/exec/HashTable.cpp [hunk 24]
```diff
@@ -181,90 +180,95 @@ class ProbeState {
...
-  void eraseHit(uint8_t* tags, int64_t& numTombstones) {
+  template <typename Table>
+  void eraseHit(Table& table, int64_t& numTombstones) {
     const auto kEmptyGroup = BaseHashTable::TagVector::broadcast(kEmptyTag);
     const bool hasEmptyGroup =
         simd::toBitMask(tagsInTable_ == kEmptyGroup) != 0;
 
-    BaseHashTable::storeTag(
-        tags, tagIndex_ + indexInTags_, hasEmptyGroup ? 0 : kTombstoneTag);
+    if (!hasEmptyGroup) {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:255]: Remove this?

File: velox/exec/HashTable.cpp [hunk 25]
```diff
@@ -288,10 +292,15 @@ void HashTable<ignoreNullKeys>::storeRowPointer(
...
     char* row) {
-  if (hashMode_ != HashMode::kArray) {
-    tags_[index] = hashTag(hash);
+  if (hashMode_ == HashMode::kArray) {
+    reinterpret_cast<char**>(table_)[index] = row;
+    return;
   }
-  table_[index] = row;
+  int64_t bucketOffset = index & ~(kBucketSize - 1);
+  auto bucket = bucketAt(bucketOffset);
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:300]:
```
nit: s/auto bucket/auto* bucket/
```

File: velox/exec/HashTable.cpp [hunk 26]
```diff
@@ -288,10 +292,15 @@ void HashTable<ignoreNullKeys>::storeRowPointer(
...
     uint64_t hash,
     char* row) {
-  if (hashMode_ != HashMode::kArray) {
-    tags_[index] = hashTag(hash);
+  if (hashMode_ == HashMode::kArray) {
+    reinterpret_cast<char**>(table_)[index] = row;
+    return;
   }
-  table_[index] = row;
+  int64_t bucketOffset = index & ~(kBucketSize - 1);
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:299]: Shall we call bucketOffset here?

File: velox/exec/HashTable.cpp [hunk 27]
```diff
@@ -538,30 +583,72 @@ void HashTable<ignoreNullKeys>::joinProbe(HashLookup& lookup) {
...
+        auto index = hashes[row];
+        DCHECK(index < capacity_);
+        hits[row] = table_[index]; // NOLINT
+      }
+    }
+  }
+  for (; i < numRows; ++i) {
+    auto row = rows[i];
+    auto index = hashes[row];
+    DCHECK(index < capacity_);
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:647]:
```
VELOX_DCHECK_LT
```

File: velox/exec/HashTable.cpp [hunk 28]
```diff
@@ -1743,23 +1859,26 @@ void HashTable<ignoreNullKeys>::checkConsistency() const {
...
+        ++numEmpty;
+        continue;
+      }
     }
   }
   VELOX_CHECK_EQ(
       numEmpty + numTombstone + numDistinct_,
       capacity_,
       "capacity: {}, numEmpty: {}, numTombstone: {}, numDistinct: {}",
       capacity_,
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:1879]: Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@oerling thanks for the update. Code now looks much cleaner!

File: velox/exec/HashTable.h [hunk 29]
```diff
@@ -15,12 +15,14 @@
...
 #pragma once
 
+#include "velox/common/base/Portability.h"
 #include "velox/common/memory/MemoryAllocator.h"
 #include "velox/exec/Operator.h"
 #include "velox/exec/RowContainer.h"
 #include "velox/exec/VectorHasher.h"
 
 namespace facebook::velox::exec {
+#define VELOX_ENABLE_INT64_BUILD_PARTITION_BOUND
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:25]: Why need this?

File: velox/exec/HashTable.cpp [hunk 30]
```diff
@@ -538,30 +583,72 @@ void HashTable<ignoreNullKeys>::joinProbe(HashLookup& lookup) {
...
+      simd::gather(
+          reinterpret_cast<const int64_t*>(table_),
+          reinterpret_cast<const int64_t*>(hashes + firstRow + kBatchSize))
+          .store_unaligned(
+              reinterpret_cast<int64_t*>(hits) + firstRow + kBatchSize);
+    } else {
+      for (auto j = i; j < i + kStep; ++j) {
+        auto row = rows[j];
+        auto index = hashes[row];
+        DCHECK(index < capacity_);
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:639]:
```
VELOX_DCHECK_LT
```

**Review by xiaoxmeng — APPROVED** [review]
@oerling LGTM. Thanks!

**Comment by skadilover** [comment]
> Experimentally, we do not see much difference, in
any case not enough to maintain two hash table layouts. 

@oerling  I found that there are 20% performance gap for 5%-hits-benchmark. Is it excepted? or Maybe 20% is enough to leave them seperately?

**Pull request closed without merging**